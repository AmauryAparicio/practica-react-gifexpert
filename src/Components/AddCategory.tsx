import React, {
  useState,
  ChangeEvent,
  FunctionComponent,
  Dispatch,
  SetStateAction,
} from "react";

import { func } from "prop-types";

export const AddCategory: FunctionComponent<{
  setCategories: Dispatch<SetStateAction<string[]>>;
}> = ({ setCategories }): JSX.Element => {
  const [inputValue, setInputValue] = useState<string>("");

  const handleInputChange = (e: ChangeEvent<HTMLInputElement>): void => {
    setInputValue(e.target.value);
  };

  const handleSubmit = (e: ChangeEvent<HTMLFormElement>): void => {
    e.preventDefault();

    if (inputValue.trim().length > 2) {
      setCategories((cats: Array<string>) => [inputValue, ...cats]);
      setInputValue("");
    }
  };

  return (
    <form onSubmit={handleSubmit}>
      <input type="text" value={inputValue} onChange={handleInputChange} />
    </form>
  );
};

AddCategory.propTypes = {
  setCategories: func.isRequired,
};
