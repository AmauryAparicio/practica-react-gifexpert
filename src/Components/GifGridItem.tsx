import React, { FunctionComponent } from "react";
import { iImg } from "./../Interfaces";
import { string } from "prop-types";

export const GifGridItem: FunctionComponent<iImg> = ({ id, title, url }) => {
  return (
    <div className="card animate__animated animate__bounceInLeft">
      <img src={url} alt={title} />
      <p>{title}</p>
    </div>
  );
};

GifGridItem.propTypes = {
  title: string.isRequired,
  url: string.isRequired,
};
