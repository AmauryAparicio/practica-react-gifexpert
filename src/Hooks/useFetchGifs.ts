import { useState, useEffect } from "react";
import { getGifs } from "./../Helpers/getGifs";
import { iState } from "../Interfaces";

export const useFetchGifs = (category: string): iState => {
  const [state, setState] = useState<iState>({
    data: [],
    loading: true,
  });

  useEffect(() => {
    getGifs(category).then((imgs) => {
      setState({
        data: imgs,
        loading: false,
      });
    });
  }, [category]);

  return state;
};
