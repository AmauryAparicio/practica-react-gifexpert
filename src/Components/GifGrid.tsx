import React, { FunctionComponent } from "react";
import { iGif } from "../Interfaces";
import { useFetchGifs } from "./../Hooks/useFetchGifs";
import { GifGridItem } from "./GifGridItem";
import { string } from "prop-types";

export const GifGrid: FunctionComponent<{ category: string }> = ({
  category,
}) => {
  const { data: images, loading } = useFetchGifs(category);

  return (
    <>
      <h3>{category}</h3>
      {loading && <p>Loading...</p>}
      <div className="card-grid">
        {images.map((img: iGif) => (
          <GifGridItem key={img.id} {...img} />
        ))}
      </div>
    </>
  );
};

GifGrid.propTypes = {
  category: string.isRequired,
};
