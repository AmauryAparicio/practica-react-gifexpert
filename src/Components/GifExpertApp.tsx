import React, { FunctionComponent, useState } from "react";
import { AddCategory } from "./AddCategory";
import { GifGrid } from "./GifGrid";

export const GifExpertApp: FunctionComponent = (): JSX.Element => {
  const [categories, setCategories] = useState<string[]>(["One Punch Man"]);

  return (
    <>
      <h2>GifExpertApp</h2>
      <AddCategory setCategories={setCategories} />
      <hr />

      <ol>
        {categories.map((category: string) => (
          <GifGrid key={category} category={category} />
        ))}
      </ol>
    </>
  );
};
