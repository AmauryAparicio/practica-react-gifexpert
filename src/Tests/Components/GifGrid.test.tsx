import { shallow } from "enzyme";
import { GifGrid } from "../../Components/GifGrid";
import React from "react";
import { useFetchGifs } from "./../../Hooks/useFetchGifs";
import "@testing-library/jest-dom";
jest.mock("./../../Hooks/useFetchGifs");

describe("Pruebas en el <GifGrid/>", () => {
  const category: string = "Testing";

  test("debe mostrarse correctamente", () => {
    useFetchGifs.mockReturnValue({
      data: [],
      loading: true,
    });

    const wrapper = shallow(<GifGrid category={category} />);

    expect(wrapper).toMatchSnapshot();
  });

  test("debe de mostrar items cuando se cargan imágenes useFetchGifs", () => {
    const gifs: Array<{ id: string; url: string; title: string }> = [
      {
        id: "ABC",
        url: "https://google.com",
        title: "esto es un título",
      },
      {
        id: "ABCDE",
        url: "https://facebook.com",
        title: "esto es un título 2",
      },
    ];

    useFetchGifs.mockReturnValue({
      data: gifs,
      loading: false,
    });

    const wrapper = shallow(<GifGrid category={category} />);

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find("p").exists()).toBe(false);
    expect(wrapper.find("GifGridItem").length).toBe(gifs.length);
  });
});
