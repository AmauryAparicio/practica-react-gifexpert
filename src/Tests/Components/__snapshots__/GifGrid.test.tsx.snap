// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el <GifGrid/> debe de mostrar items cuando se cargan imágenes useFetchGifs 1`] = `
<Fragment>
  <h3>
    Testing
  </h3>
  <div
    className="card-grid"
  >
    <GifGridItem
      id="ABC"
      key="ABC"
      title="esto es un título"
      url="https://google.com"
    />
    <GifGridItem
      id="ABCDE"
      key="ABCDE"
      title="esto es un título 2"
      url="https://facebook.com"
    />
  </div>
</Fragment>
`;

exports[`Pruebas en el <GifGrid/> debe mostrarse correctamente 1`] = `
<Fragment>
  <h3>
    Testing
  </h3>
  <p>
    Loading...
  </p>
  <div
    className="card-grid"
  />
</Fragment>
`;
